core/src/main/java/org/infinispan/commands/write/ComputeCommand.java

```diff
@@ -0,0 +1,223 @@
+import org.infinispan.container.entries.MVCCEntry;
+import org.infinispan.context.InvocationContext;
+import org.infinispan.context.impl.FlagBitSets;
+import org.infinispan.factories.ComponentRegistry;
+import org.infinispan.metadata.Metadata;
+import org.infinispan.metadata.Metadatas;
+import org.infinispan.notifications.cachelistener.CacheNotifier;
+
+public class ComputeCommand extends AbstractDataWriteCommand implements MetadataAwareCommand {
+
+   public static final int COMMAND_ID = 68;
+
+   private BiFunction remappingBiFunction;
+   private Metadata metadata;
+   private CacheNotifier<Object, Object> notifier;
+   private boolean computeIfPresent;
+
+   public ComputeCommand() {
+   }
+
+   public ComputeCommand(Object key,
+                         BiFunction remappingBiFunction,
+                         boolean computeIfPresent,
+                         long flagsBitSet,
+                         CommandInvocationId commandInvocationId,
+                         Metadata metadata,
+                         CacheNotifier notifier,
+                         ComponentRegistry componentRegistry) {
+
+      super(key, flagsBitSet, commandInvocationId);
+      this.remappingBiFunction = remappingBiFunction;
+      this.computeIfPresent = computeIfPresent;
+      this.metadata = metadata;
+      this.notifier = notifier;
+      componentRegistry.wireDependencies(this.remappingBiFunction);
+   }
+
+   public boolean isComputeIfPresent() {
+      return computeIfPresent;
+   }
+
+   public void setComputeIfPresent(boolean computeIfPresent) {
+      this.computeIfPresent = computeIfPresent;
+   }
+
+   public void init(CacheNotifier notifier, ComponentRegistry componentRegistry) {
+      //noinspection unchecked
+      this.notifier = notifier;
+      componentRegistry.wireDependencies(remappingBiFunction);
+   }
+
+   @Override
+   public Metadata getMetadata() {
+      return metadata;
+   }
+
+   @Override
+   public void setMetadata(Metadata metadata) {
+      this.metadata = metadata;
+   }
+
+   @Override
+   public boolean isSuccessful() {
+      return true;
+   }
+
+   @Override
+   public boolean isConditional() {
+      return computeIfPresent;
+   }
+
+   @Override
+   public ValueMatcher getValueMatcher() {
+      return ValueMatcher.MATCH_ALWAYS;
+   }
+
+   @Override
+   public void setValueMatcher(ValueMatcher valueMatcher) {
+      //implementation not needed
+   }
+
+   @Override
+   public void updateStatusFromRemoteResponse(Object remoteResponse) {
+      // implementation not needed
+   }
+
+   @Override
+   public Object perform(InvocationContext ctx) throws Throwable {
+      MVCCEntry<Object, Object> e = (MVCCEntry) ctx.lookupEntry(key);
+
+      if (e == null) {
+         throw new IllegalStateException("Not wrapped");
+      }
+
+      Object oldValue = e.getValue();
+
+      if (oldValue != null) {
+         Object newValue = remappingBiFunction.apply(key, oldValue);
+         if (newValue != null) {
```

Comment:
If `newValue == oldValue` I think we shouldn't call the notifier nor update the entry.